src/libraries/Common/src/System/Security/Cryptography/KeyFormatHelper.Encrypted.cs

```diff
@@ -0,0 +1,393 @@
+                ReadOnlySpan<byte>.Empty,
+                pkcs8Writer,
+                pbeParameters);
+        }
+
+        internal static AsnWriter WriteEncryptedPkcs8(
+            ReadOnlySpan<byte> passwordBytes,
+            AsnWriter pkcs8Writer,
+            PbeParameters pbeParameters)
+        {
+            return WriteEncryptedPkcs8(
+                ReadOnlySpan<char>.Empty,
+                passwordBytes,
+                pkcs8Writer,
+                pbeParameters);
+        }
+
+        private static AsnWriter WriteEncryptedPkcs8(
+            ReadOnlySpan<char> password,
+            ReadOnlySpan<byte> passwordBytes,
+            AsnWriter pkcs8Writer,
+            PbeParameters pbeParameters)
+        {
+            PasswordBasedEncryption.InitiateEncryption(
+                pbeParameters,
+                out SymmetricAlgorithm cipher,
+                out string hmacOid,
+                out string encryptionAlgorithmOid,
+                out bool isPkcs12);
+
+            byte[]? encryptedRent = null;
+            Span<byte> encryptedSpan = default;
+            AsnWriter? writer = null;
+
+            try
+            {
+                Debug.Assert(cipher.BlockSize <= 128, $"Encountered unexpected block size: {cipher.BlockSize}");
+                Span<byte> iv = stackalloc byte[cipher.BlockSize / 8];
+                Span<byte> salt = stackalloc byte[16];
+
+                // We need at least one block size beyond the input data size.
+                encryptedRent = CryptoPool.Rent(
+                    checked(pkcs8Writer.GetEncodedLength() + (cipher.BlockSize / 8)));
+
+                RandomNumberGenerator.Fill(salt);
+
+                int written = PasswordBasedEncryption.Encrypt(
+                    password,
+                    passwordBytes,
+                    cipher,
+                    isPkcs12,
+                    pkcs8Writer,
+                    pbeParameters,
+                    salt,
+                    encryptedRent,
+                    iv);
+
+                encryptedSpan = encryptedRent.AsSpan(0, written);
+
+                writer = new AsnWriter(AsnEncodingRules.DER);
+
+                // PKCS8 EncryptedPrivateKeyInfo
+                writer.PushSequence();
+
+                // EncryptedPrivateKeyInfo.encryptionAlgorithm
+                PasswordBasedEncryption.WritePbeAlgorithmIdentifier(
+                    writer,
+                    isPkcs12,
+                    encryptionAlgorithmOid,
+                    salt,
+                    pbeParameters.IterationCount,
+                    hmacOid,
+                    iv);
+
+                // encryptedData
+                writer.WriteOctetString(encryptedSpan);
+                writer.PopSequence();
+
+                return writer;
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(encryptedSpan);
+                CryptoPool.Return(encryptedRent!, clearSize: 0);
```

Comment:
_encryptedRent_ may be null if an exception is thrown on line 203 - 204. I'd recommend moving lines 198 - 206 to be before the _try_ block, which obviates this issue; or putting a null guard around this.

Reply:
Since this is just a partial split-out with no changes, and there's a clock running for the preview snap, I'll tackle this across all the algorithms as a followup.